src/axom/core/FlatMap.hpp

```diff
@@ -0,0 +1,771 @@
+
+  /*!
+   * \brief Remove a key-value pair from the FlatMap, specified by key.
+   *
+   *  If the key doesn't exist in the FlatMap, does nothing.
+   *
+   * \param key the key to remove
+   *
+   * \return 1 if an entry was removed, 0 otherwise
+   */
+  IndexType erase(const KeyType& key)
+  {
+    const_iterator it = find(key);
+    if(it != end())
+    {
+      erase(it);
+      return 1;
+    }
+    return 0;
+  }
+
+  /*!
+   * \brief Returns the number of buckets allocated in the FlatMap.
+   *
+   *  The maximum number of elements that can be stored in the FlatMap without
+   *  resizing and rehashing is bucket_count() * max_load_factor().
+   */
+  IndexType bucket_count() const { return m_buckets.size(); }
+
+  /*!
+   * \brief Returns the current load factor of the FlatMap.
+   */
+  double load_factor() const { return ((double)m_loadCount) / bucket_count(); }
+
+  /*!
+   * \brief Returns the maximum load factor of the FlatMap.
+   */
+  double max_load_factor() const { return MAX_LOAD_FACTOR; }
+
+  /*!
+   * \brief Explicitly rehash the FlatMap with a given number of buckets.
+   *
+   * \param count the minimum number of buckets to allocate for the rehash
+   */
+  void rehash(IndexType count)
+  {
+    FlatMap rehashed(m_size,
+                     std::make_move_iterator(begin()),
+                     std::make_move_iterator(end()),
+                     count);
+    this->swap(rehashed);
+  }
+
+  /*!
+   * \brief Reallocate and rehash the FlatMap, such that up to the specified
+   *  number of elements may be inserted without a rehash.
+   *
+   * \param count the number of elements to fit without a rehash
+   */
+  void reserve(IndexType count) { rehash(std::ceil(count / MAX_LOAD_FACTOR)); }
+
+private:
+  template <typename InputIt>
+  FlatMap(IndexType num_elems, InputIt first, InputIt last, IndexType bucket_count);
+
+  std::pair<iterator, bool> getEmplacePos(const KeyType& key);
+
+  template <typename... Args>
+  void emplaceImpl(const std::pair<iterator, bool>& pos,
+                   bool assign_on_existence,
+                   Args&&... args);
+
+  constexpr static IndexType MIN_NUM_BUCKETS {29};
+
+  IndexType m_numGroups2;  // Number of groups of 15 buckets, expressed as a power of 2
+  IndexType m_size;
+  axom::Array<detail::flat_map::GroupBucket> m_metadata;
+
+  // Storage details:
+  struct alignas(KeyValuePair) PairStorage
+  {
+    unsigned char data[sizeof(KeyValuePair)];
+
+    const KeyValuePair& get() const
+    {
+      return *(reinterpret_cast<const KeyValuePair*>(&data));
+    }
+
+    KeyValuePair& get() { return *(reinterpret_cast<KeyValuePair*>(&data)); }
+  };
+
+  axom::Array<PairStorage> m_buckets;
+
+  // Boost flat_unordered_map uses a fixed load factor.
+  constexpr static double MAX_LOAD_FACTOR = 0.875;
+  std::uint64_t m_loadCount;
+};
+
+template <typename KeyType, typename ValueType, typename Hash>
+template <bool Const>
+class FlatMap<KeyType, ValueType, Hash>::IteratorImpl
```

Comment:
We can't implement `FlatMap::iterator::operator++()` in constant-time, since we have to iterate through an indeterminate number of buckets in order to find the next occupied bin. Since `IteratorBase` models a `RandomAccessIterator`, I think that precludes us from implementing `FlatMap::Iterator` in terms of `IteratorBase`.